### ISSUE TITLE
UnnecessaryAbstractClass false positive

### ISSUE THREAD

**Issue by TWiStErRob** [detekt/detekt#4753]
## Expected Behavior
Don't report following code:
```
abstract class Fake : Base() // UnnecessaryAbstractClass
abstract class Base {
	open val order: Int get() = 0
	abstract fun String.foo()
}
```

## Observed Behavior

> "An abstract class without a concrete member can be refactored to an interface."

No, it can't:
 * `Fake` interface couldn't extend a class.
 * it has concrete members inherited from `Base`
 * I get a detection even if I add an actual class using the "abstractness":
```kotlin
class Foo : Fake() {
	override fun String.foo() = TODO("not implemented")
}
```
Note: `String.` might be a red herring, you can ignore that part, just wanted to report what I had. More minimal repro might exist.

## Steps to Reproduce

Just copy above code into a test or code checked by Detekt 1.20.0.

## Context
Real context is creating fakes for a test:
![image](https://user-images.githubusercontent.com/2906988/164911355-ca1cedf3-2ea0-4b1b-806e-c660f88e700d.png)

The abstract method is never invoked in the test, so there's no need to implement it, hence using `abstract class`. The class must be unique because it's `::class` object is used.

## Your Environment

* Version of detekt used: 1.20.0
* Version of Gradle used (if applicable): N/A
* Gradle scan link (add `--scan` option when running the gradle task): N/A
* Operating System and version: all
* Link to your project (if it's a public repository): https://github.com/TWiStErRob/net.twisterrob.cinema/pull/114/commits/0a43c8e914bcd46fd9db4a80baeba0470a7e07df in https://github.com/TWiStErRob/net.twisterrob.cinema/pull/114

**Comment by BraisGabin** [comment]
Isn't this an edge case? I mean, in general a class that only extends another class and does nothing is unnecessary, you could use the base class. In this case the code uses the type but that seems to me like an edge case.

I'm not against this change but I'm worried that by allowing this we will add a lot of false negatives.

**Comment by TWiStErRob** [comment]
If you build an expression hierarchy, or similar, you might introduce an intermediate base class for grouping part of the class hierarchy.

**Comment by dawidhyzy-ifolor** [comment]
We face the same issue when wrapping a third party library classes so it's not an edge case.

**Comment by BraisGabin** [comment]
I agree, maybe we could add a configuration here but I don't think that it could be even needed. Do you have bandwith to provide a PR?

**Comment by TWiStErRob** [comment]
Me, not for a month probably.

@dawidhyzy-ifolor can you please share your use case, it might help with unit tests.

**Comment by dawidhyzy-ifolor** [comment]
Sure!

We have `DataSource` defined in the `domain` layer

```
abstract class PhotosMediaItemPagedDataSource : PagingSource<NextPageToken, MediaItem>() {

    interface Factory {
        fun create(albumId: AlbumId?, pageSize: Int?): PhotosMediaItemPagedDataSource
    }
}
```

As `androidx.paging.PagingSource` is an abstract class we can't refactor `PhotosMediaItemPagedDataSource` to an interface - interface can't extend abstract class.

Please, let me know if you need more information.

**Comment by chao2zhang** [comment]
I think the rule of thumb is because `Interface can't extend abstract class`. I agree that we should at least provide a configuration to this rule.

**Comment by gouri-panda** [comment]
@BraisGabin May i take this?

**Comment by BraisGabin** [comment]
Go ahead

**Comment by BraisGabin** [comment]
I'm fine with or without configuration. I suppose that in that case without is better because it keeps the api cleaner but that's your call. I think that no one here have a hard opinion about this topic.

**Comment by TWiStErRob** [comment]
Ask yourself if there's a possibility for the case you're flagging. It it possible to get code that passes/fails the rule? And are both of them valid engineering choices? (i.e. due to decisions on preferences or restricted by constraints)

**Comment by EdwarDDay** [comment]
This issue is closed, but it still happens in 1.21.0 (gradle plugin). Am I doing something wrong?
Example run: https://github.com/EdwarDDay/detekt-abstract-class-unnecessary/runs/7410220388?check_suite_focus=true
Example code: 
```kotlin
abstract class Foo {
    abstract val i: Int
    val j = 1
}

abstract class Bar : Foo() // UnnecessaryAbstractClass issue
```

**Comment by TWiStErRob** [comment]
That is very strange since that almost exact example is in a unit test:

https://github.com/detekt/detekt/pull/4399/files#diff-28879f691cfd6348ba8a90a837260b8c317e548295c9c400b1199de7bafdc8f4R75-R85

**Comment by TWiStErRob** [comment]
I can confirm that the fix did not fix the reported problem:
https://github.com/TWiStErRob/net.twisterrob.cinema/pull/173/files

**Comment by TWiStErRob** [comment]
@BraisGabin ca  you please reopen?

**Comment by atulgpt** [comment]
Hi, @EdwarDDay the TC given by you is passing when you run detekt with type resolution on but if you run w/o type resolution then it is failing. Can you confirm in which mode you are running the detekt?
Hi @TWiStErRob can you confirm in the PR you run detekt /w type resolution or w/o it? cc @BraisGabin

**Comment by EdwarDDay** [comment]
Yes, I just ran the `detekt` task and not `detektMain`. When I read the [docs](https://detekt.dev/docs/gettingstarted/type-resolution/) right, this means, that the task runs w/o type resolution.

**Comment by TWiStErRob** [comment]
@atulgpt If you meant https://github.com/detekt/detekt/issues/4753#issuecomment-1191312692, the linked PR https://github.com/TWiStErRob/net.twisterrob.cinema/pull/173 has a [commit](https://github.com/TWiStErRob/net.twisterrob.cinema/pull/173/commits/2b01c6258027f0adac3435bf91e345376fac468a) whose [build](https://github.com/TWiStErRob/net.twisterrob.cinema/runs/7447240598) found this. Navigating the build you can find the [workflow file](https://github.com/TWiStErRob/net.twisterrob.cinema/actions/runs/2711038212/workflow#L68) which refers to [this hacky w/o type resolution task](https://github.com/TWiStErRob/net.twisterrob.cinema/blob/2b01c6258027f0adac3435bf91e345376fac468a/Heroku/build.gradle.kts#L160-L169) (since then I've moved on to [per-module type resolution](https://github.com/TWiStErRob/net.twisterrob.cinema/pull/235))

**Comment by atulgpt** [comment]
Hi @TWiStErRob yup talking bout https://github.com/detekt/detekt/issues/4753#issuecomment-1191312692
if this is resolved we can close this issue
